### ISSUE TITLE
[FR] Set micro-stepping (on the fly) on TMC Drivers [NVI]

### ISSUE THREAD

**Issue by qwewer0** [MarlinFirmware/Marlin#19499]
### Description

It would be useful to be able to use M350 - Set micro-stepping with TMC drivers or to have a dedicated TMC command to set micro-stepping. (e.g. SKR Mini E3 v1.2, TMC 2209)

### Feature Workflow

1. Send M350 ... (or a TMC specific one) to the printer (e.g. M350 E1)
2. Get OK response, instead of Unknown Command: M350 ...
3. E stepper is now run with 1 micro stepping

Edit: [NVI] = Not Very Important

**Comment by thisiskeithb** [comment]
No. `M350` is for drivers that have digital microstepping pins which the SKR mini boards do not have. https://marlinfw.org/docs/gcode/M350.html

<hr>

This Issue Queue is for Marlin bug reports and development-related issues, and we prefer not to handle user-support questions here. (As noted on [this page](https://github.com/MarlinFirmware/Marlin/blob/2.0.x/.github/contributing.md#i-dont-want-to-read-this-whole-thing-i-just-have-a-question).) For best results getting help with configuration and troubleshooting, please use the following resources:

- MarlinFW.org [Marlin Documentation](http://marlinfw.org)
- RepRap.org [Marlin Forum](http://forums.reprap.org/list.php?415)
- [Tom's 3D Forums](https://forum.toms3d.org/)
- Facebook Group ["Marlin Firmware"](https://www.facebook.com/groups/1049718498464482/)
- Facebook Group ["Marlin Firmware for 3D Printers"](https://www.facebook.com/groups/3Dtechtalk/)
- [Marlin Configuration](https://www.youtube.com/results?search_query=marlin+configuration) on YouTube
- Marlin Discord server. Join link: https://discord.gg/n5NJ59y

After seeking help from the community, if the consensus points to a bug in Marlin, then you should post a [bug report](https://github.com/MarlinFirmware/Marlin/issues/new/choose).

**Comment by sjasonsmith** [comment]
@thisiskeithb why close this? Seems perfectly valid as a feature request. We had already discussed it on Discord.

**Comment by thisiskeithb** [comment]
> @thisiskeithb why close this? Seems perfectly valid as a feature request. We had already discussed it on Discord.

This seemed like a question about `M350`'s use and not a feature request.

Since you're aware/handling it, I'll reopen it.

**Comment by qwewer0** [comment]
Changed the description to be more like a FR and not a question.

**Comment by thisiskeithb** [comment]
Following the [Feature Request template](https://raw.githubusercontent.com/MarlinFirmware/Marlin/2.0.x/.github/ISSUE_TEMPLATE/feature_request.md) really helps 🙂

**Comment by sjasonsmith** [comment]
The M350 command on RepRap seems to be documented to allow up to 256 microsteps. It doesn't clarify whether it works with TMC drivers, but to support 256 microsteps that seems incredibly likely.
https://www.reprap.org/wiki/G-code#M350:_Set_microstepping_mode

Yep, looks like they are using M350 to set both microstepping and interpolation on TMC drivers:
https://github.com/Duet3D/RepRapFirmware/blob/bcb169fbf6fc8457afb272cd8ddf3d6b76d58795/src/Movement/StepperDrivers/TMC22xx.cpp#L568

I imagine to be useful it would also have to store settings to EEPROM.

**Comment by qwewer0** [comment]
<s>@sjasonsmith So, that means it works for TMC drivers already, and can close this FR, or while it works, it still should be able to make it work on boards that has TMC drivers without digital micro-stepping pins?</s>

**Comment by sjasonsmith** [comment]
I was just looking at what RepRap firmware does with the command. I wasn’t referring to Marlin code at all. There is some desire to  use the same commands across different firmwares where possible. RepRap configuring this through M350 makes it easier to justify adding it to Marlin.

I think M350 would be the place to do it, the remaining question is whether making it runtime configurable is a good idea. Every constant which becomes configurable makes the math a little harder while printing, which may be noticeable on slower controllers.

**Comment by teemuatlut** [comment]
I don't think it would be used during printing. But I question if this command is something that would be used more than just during initial setup and I think this is just not something that is "tuned". It's basically a set it and forget it type of configuration item.
Then if you want to have the command for setting this live, then you'll likely also want it to be saved in EEPROM and overall I just don't see the point.

**Comment by qwewer0** [comment]
One use case I can think of right now, is a "cold pull" gcode. If the extruder isn't strong enough on 16 micros-tepping to pull the filament out, then switching it to full stepping would result stronger extrusion, and thus would be able to pull the filament out. Assuming this is how the stepper motor torque changes relation to the micro-stepping.

**Comment by teemuatlut** [comment]
Microstepping doesn't decrease torque.
While you're taking finer steps, you're also taking comparatively more steps and the end result is the same delta between rotor and stator, thus resulting in the same applied torque.

**Comment by qwewer0** [comment]
So you are saying that micro-stepping doesn't decreases the rotation torque, only the holding torque on a micro step? And so it would pull the filament with the exact same amount of force?
Edit:
[Microstepping](https://www.faulhaber.com/fileadmin/user_upload_global/support/MC_Support/Motors/AppNotes/Faulhaber_AN015_EN.pdf)
[Microstepping myths](https://www.machinedesign.com/archive/article/21812154/microstepping-myths)
Not sure if I read these two correctly...

**Comment by thegwydd** [comment]
I have an Ender 3 with Creality 4.2.7 motherboard and I've just installed a Flex3Drive extruder.
I'd like to know if on my board it's possible to adjust microsteps to 8x.
Thank you

**Comment by mihtjel** [comment]
Probably gonna look into implementing this tomorrow or Sunday as I want to try out some different settings and can't be bothered recompiling and flashing just for that.

(Yes, yes. I am going to implement this feature _because I'm lazy_, despite it doubtless being much more work.)

**Comment by mikhail11111** [comment]
Is there any progress here?

**Comment by ellensp** [comment]
You do realize with most tmc's they are already set to   #define INTERPOLATE      true
So the only effect that changing the micro stepping does is to change the loading on the micro controller?

**Comment by DiskDoctor** [comment]
Is M350 enabled for >16 yet? Looking to find a way to identify TMC drivers via software vs tearing apart printers and heat sinks. 
If M350 S256 (for example) 
can be used, TMC's should respond, A4988's should report error. Is this possible?

**Comment by ellensp** [comment]
@DiskDoctor take a look at macro AXIS_IS_TMC(A)

**Comment by DiskDoctor** [comment]
@ellensp  Thanks. Could you kindly point me to the macros? I've used search but no results. Discord, another link?

**Comment by thisiskeithb** [comment]
> I've used search but no results.

Run a global search in VSCode:
<img width="289" alt="image" src="https://user-images.githubusercontent.com/13375512/166319579-1c7c8b25-4d6e-430a-8b23-169ee1acb710.png">

**Comment by mikhail11111** [comment]
> You do realize with most tmc's they are already set to #define INTERPOLATE true So the only effect that changing the micro stepping does is to change the loading on the micro controller?

Right, hence the maximum speed attainable
